src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,539 @@
+                        throw new InvalidOperationException(Strings.InvalidOperation_ConcurrentOperationsNotSupported);
+                    }
+                    collisionCount++;
+                }
+
+                return ref AddKey(key, bucketIndex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ref TValue AddKey(TKey key, int bucketIndex)
+        {
+            Entry[] entries = _entries;
+            int entryIndex;
+            if (_freeList != -1)
+            {
+                entryIndex = _freeList;
+                _freeList = -3 - entries[_freeList].next;
+            }
+            else
+            {
+                if (Count == entries.Length || entries.Length == 1)
+                {
+                    entries = Resize();
+                    bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                    // entry indexes were not changed by Resize
+                }
+                entryIndex = Count;
+            }
+
+            entries[entryIndex].key = key;
+            entries[entryIndex].next = _buckets[bucketIndex] - 1;
+            _buckets[bucketIndex] = entryIndex + 1;
+            Count++;
+            return ref entries[entryIndex].value;
+        }
+
+        private Entry[] Resize()
+        {
+            int count = Count;
+            int newSize = HashHelpers.ExpandPrime(count);
+            var entries = new Entry[newSize];
+            Array.Copy(_entries, 0, entries, 0, count);
+
+            var newBuckets = new int[newSize];
+            while (count-- > 0)
+            {
+                int bucketIndex = (int)(GetHashCode(entries[count].key) % (uint)newBuckets.Length);
+                entries[count].next = newBuckets[bucketIndex] - 1;
+                newBuckets[bucketIndex] = count + 1;
+            }
+
+            _buckets = newBuckets;
+            _entries = entries;
+
+            return entries;
+        }
+
+        public KeyEnumerable Keys => new KeyEnumerable(this);
+
+        public ValueEnumerable Values => new ValueEnumerable(this);
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
+        {
+            Entry[] entries = _entries;
+            int i = 0;
+            int count = Count;
+            while (count > 0)
+            {
+                if (entries[i].next > -2)
+                {
+                    count--;
+                    array[index++] = new KeyValuePair<TKey, TValue>(
+                        entries[i].key,
+                        entries[i].value);
+                }
+                i++;
+            }
+        }
+
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
```

Comment:
We now have 3 public nested structs named Enumerator. Can they be internal, so they don't need documenting? We only hand out the IEnumerable interfaces, and those can be cast to IDisposable if needed. I realize Dictionary has them public too, but I don't know why. Perhaps because they thought it was ugly for users to put them in variables with interface type, but they are almost always used directly in foreach.

Reply:
They should be exposed via the strongly typed non-allocating struct enumerators e.g. `Enumerator GetEnumerator()`

Reply:
Aha - that's why they need to be public. Better now?